server/coordinator/procedure/manager_impl.go

```diff
@@ -0,0 +1,182 @@
+		return nil
+	}
+	m.procedureQueue = make(chan Procedure, queueSize)
+	go m.startProcedureWorker(ctx, m.procedureQueue)
+	err := m.retryAll(ctx)
+	if err != nil {
+		return errors.WithMessage(err, "retry all running procedure failed")
+	}
+	return nil
+}
+
+func (m *ManagerImpl) Stop(ctx context.Context) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	close(m.procedureQueue)
+	for _, procedure := range m.procedures {
+		if procedure.State() == StateRunning {
+			err := procedure.Cancel(ctx)
+			log.Error("cancel procedure failed", zap.Error(err), zap.Uint64("procedureID", procedure.ID()))
+			// TODO: consider whether a single procedure cancel failed should return directly.
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *ManagerImpl) Submit(_ context.Context, procedure Procedure) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.procedures = append(m.procedures, procedure)
+	return nil
+}
+
+func (m *ManagerImpl) Cancel(ctx context.Context, procedureID uint64) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	for _, procedure := range m.procedures {
+		if procedure.ID() == procedureID {
+			err := procedure.Cancel(ctx)
+			if err != nil {
+				return errors.WithMessagef(err, "cancel procedure failed, procedureID:%d", procedureID)
+			}
+			return nil
+		}
+	}
+	log.Error("procedure not found", zap.Uint64("procedureID", procedureID))
+	return ErrProcedureNotFound
+}
+
+func (m *ManagerImpl) ListRunningProcedure(_ context.Context) ([]*Info, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	procedureInfos := make([]*Info, 0)
+	for _, procedure := range m.procedures {
+		if procedure.State() == StateRunning {
+			procedureInfos = append(procedureInfos, &Info{
+				ID:    procedure.ID(),
+				Typ:   procedure.Typ(),
+				State: procedure.State(),
+			})
+		}
+	}
+	return procedureInfos, nil
+}
+
+func NewManagerImpl(client *clientv3.Client, rootPath string) (Manager, error) {
+	manager := &ManagerImpl{
+		storage:  NewEtcdStorageImpl(client, rootPath),
+		dispatch: eventdispatch.NewDispatchImpl(),
+	}
+	return manager, nil
+}
+
+func (m *ManagerImpl) retryAll(ctx context.Context) error {
+	metas, err := m.storage.List(ctx, metaListBatchSize)
+	if err != nil {
+		return errors.WithMessage(err, "storage list meta failed")
+	}
+	for _, meta := range metas {
+		if !meta.needRetry() {
+			continue
+		}
+		p := load(meta)
+		err := m.retry(ctx, p)
+		return errors.WithMessagef(err, "retry procedure failed, procedureID:%d", p.ID())
+	}
+	return nil
+}
+
+func (m *ManagerImpl) startProcedureWorker(ctx context.Context, procedures <-chan Procedure) {
+	for procedure := range procedures {
+		err := procedure.Start(ctx)
+		if err != nil {
+			log.Error("procedure start failed", zap.Error(err))
+		}
+
+		m.lock.Lock()
+		index := -1
+		for i, p := range m.procedures {
+			if p.ID() == procedure.ID() {
+				index = i
+			}
+		}
+		if index != -1 {
+			m.procedures = append(m.procedures[:index], m.procedures[index+1:]...)
+		}
+		m.lock.Unlock()
```

Comment:
How about encapsulating this block as a method of Manager called `removeProcedure`? And call this when cancel one procedure too.